app/assets/javascripts/discourse/app/lib/hashtag-autocomplete.js

```diff
@@ -42,17 +84,80 @@ function _setup($textArea, siteSettings, afterComplete) {
     template: findRawTemplate("category-tag-autocomplete"),
     key: "#",
     afterComplete,
-    transformComplete: (obj) => {
-      return obj.text;
-    },
+    transformComplete: (obj) => obj.text,
     dataSource: (term) => {
       if (term.match(/\s/)) {
         return null;
       }
       return searchCategoryTag(term, siteSettings);
     },
-    triggerRule: (textarea, opts) => {
-      return categoryHashtagTriggerRule(textarea, opts);
-    },
+    triggerRule: (textarea, opts) => hashtagTriggerRule(textarea, opts),
   });
 }
+
+let searchCache = {};
+let searchCacheTime;
+let currentSearch;
+
+function _updateSearchCache(term, results) {
+  searchCache[term] = results;
+  searchCacheTime = new Date();
+  return results;
+}
+
+function _searchGeneric(term, siteSettings, context) {
+  if (currentSearch) {
+    currentSearch.abort();
+    currentSearch = null;
+  }
+  if (new Date() - searchCacheTime > 30000) {
+    searchCache = {};
+  }
+  const cached = searchCache[term];
+  if (cached) {
+    return cached;
+  }
+
+  return new Promise((resolve) => {
+    let timeoutPromise = isTesting()
+      ? null
+      : discourseLater(() => {
+          resolve(CANCELLED_STATUS);
+        }, 5000);
+
+    if (term === "") {
+      return resolve(CANCELLED_STATUS);
+    }
+
+    const debouncedSearch = (q, ctx, resultFunc) => {
+      discourseDebounce(this, _searchRequest, q, ctx, resultFunc, INPUT_DELAY);
+    };
+
+    debouncedSearch(term, context, (result) => {
+      cancel(timeoutPromise);
+      resolve(_updateSearchCache(term, result));
+    });
+  });
+}
+
+function _searchRequest(term, context, resultFunc) {
+  currentSearch = ajax("/hashtags/search.json", {
+    data: { term, order: _sortedContextParams(context) },
+  });
+  let returnVal = CANCELLED_STATUS;
+  currentSearch
+    .then((r) => {
+      returnVal = r.results;
+      resultFunc(returnVal);
```

Comment:
I dont understand the logic here. The only place we use returnVal is after doing "returnVal = r.results", so when is `let returnVal = CANCELLED_STATUS;`going to be used, and why don't we just do `resultFunc(r.results)`  or maybe `resultFunc(r.results || CANCELLED_STATUS)` ?

Reply:
we could also probably just chain everything:

```
return ajax(...).then(...);
```

Not sure what `currentSearch = null` is used for here?

Reply:
This is all copied from the old category-tag autocomplete code here https://github.com/discourse/discourse/blob/0eab3f45c4f7fde602717e538edf7849a111bdb8/app/assets/javascripts/discourse/app/lib/category-tag-search.js . I thought maybe there was some reason it was done this way, but maybe not 🤷 I can rearrange . Will probably do `resultFunc(r.results || CANCELLED_STATUS)` as suggested

Reply:
currentSearch is aborted when we debounce the autocomplete